pyccel/ast/builtin_objects/list_functions.py

```diff
@@ -36,12 +36,27 @@ class ListAppend(PyccelInternalFunction):
     """
     __slots__ = ()
     _dtype = NativeVoid()
+    _shape = None
+    _order = None
+    _rank = 0
+    _precision = -1
+    _class_type = NativeHomogeneousList()
     name = 'append'
 
-    def __init__(self, *args) -> None:
-        super().__init__(*args)
-        if not isinstance(args[1].dtype, type(args[0].dtype)):
-            errors.report(f'Argument of type {args[1].dtype} is incompatible with list of type {args[0].dtype}',
-                        severity='fatal')
-
-
+    def __init__(self, lst_bound_arg, new_elem) -> None:
+        super().__init__(new_elem)
+        list_precision = get_final_precision(lst_bound_arg)
+        conditions = [
+            new_elem.dtype is not NativeGeneric() and
+            lst_bound_arg.dtype == new_elem.dtype and
+            list_precision == get_final_precision(new_elem) and
+            lst_bound_arg.rank - 1 == new_elem.rank and
+            (getattr(lst_bound_arg, 'order', None) == getattr(new_elem, 'order',None))
+        ]
+        is_homogeneous = lst_bound_arg.dtype is not NativeGeneric() and all(conditions)
+        if is_homogeneous:
+            inner_shape = [() if new_elem.rank == 0 else new_elem.shape]
+            new_list_shape = (lst_bound_arg.shape,) + inner_shape[0]
+            lst_bound_arg.set_init_shape(new_list_shape)
```

Comment:
The "init shape" is the shape at the initialisation. Any append statement happens after the initialisation so it doesn't make sense to modify this value